repo/blob/azure/azure_storage.go

```diff
@@ -142,6 +151,22 @@ func (az *azStorage) PutBlob(ctx context.Context, b blob.ID, data blob.Bytes, op
 		return translateError(err)
 	}
 
+	if opts.RetentionPeriod != 0 {
+		// it will fail if the retentionPeriod set by the user is lower than that on the policy.
+		retainUntilDate := clock.Now().Add(opts.RetentionPeriod).UTC()
+		mode := azblobblob.ImmutabilityPolicySetting(opts.RetentionMode)
+
+		_, err = az.service.ServiceClient().
+			NewContainerClient(az.Container).
+			NewBlobClient(az.getObjectNameString(b)).
+			SetImmutabilityPolicy(ctx, retainUntilDate, &azblobblob.SetImmutabilityPolicyOptions{
+				Mode: &mode,
+			})
+		if err != nil {
+			return errors.Wrap(err, "unable to extend retention period")
+		}
+	}
```

Comment:
Only in the case when there is no policy on the container. If there is a policy (presumably it would be recommended in kopia docs) then there would default protection also. At first I didn't include this step (just depending a the policy) but decided to add it to have some support without a policy as an option.

I created a request with Azure SDK to make it possible to set during upload but it'll take a while https://github.com/Azure/azure-sdk-for-go/issues/21620